### PR TITLE
Make timer great again

### PR DESCRIPTION
**Pull request by chrisveilleux**
#### Description
Refactor the entirety of this skill to improve readability, maintainability and robustness.  This PR is massive.  Checking out the branch and reviewing the code in its entirety would be easier than trying to browse through all the changes.

Requires [skill-date-time PR115](https://github.com/MycroftAI/skill-date-time/pull/115).

#### Type of PR
If your PR fits more than one category, there is a high chance you should submit more than one PR. Please consider this carefully before opening the PR.
_Either delete those that do not apply, or add an x between the square brackets like so: `- [x]`_
- [ ] Bugfix
- [ ] Feature implementation
- [x] Refactor of code (without functional changes)
- [ ] Documentation improvements
- [ ] Test improvements

#### Testing
VK tests are passing, with the exception of one or two expected fail scenarios.  Timer display on Mark I and Mark II should work as designed.

#### Documentation
All the new code has module, class and function level docstrings.

### CODE REVIEW

File: __init__.py [hunk 1]
```diff
@@ -11,1012 +11,863 @@
...
+        displaying the timer during these events.
+        """
+        self._stop_expiration_check()
+        if self.platform == MARK_I:
+            self._stop_timer_display()
+
+    def _resume_scheduled_events(self):
+        """Resume scheduled events that were paused during listening/speaking."""
+        self._start_timer_display()
+        if self.platform == MARK_I:
```
Comment (forslund) [inline __init__.py]: Should this line be
```python
        if len(self.active_timers > 0):
```

?

I was chatting with @emphasize about this and the timer events weren't being triggered properly. It turned out it was stopped and then not restarted. When we debugged we thought this looked a bit weird, and making the above change makes the event trigger as  expected.

**Comment by emphasize** [comment]
I think its appropriate to throw some test things in here.

the "common stop" logic (`def stop`) and the cancel intent are overlapping. "stop" has to be erased from `cancel.voc`, especially since the intent seems to be designed around cancelling non-expired timers. 
This isn't recognizeable if only one unnamed is started and cancelled before/after expiry, but otherwise

**Comment by emphasize** [comment]
[This](https://github.com/MycroftAI/mycroft-timer/blob/b67d8baf7cfc750e577518f7148322a08031e55c/__init__.py#L707) should be:
```python
def _display_timers_on_gui(self):
        """Display active timers on a device that supports the QT GUI framework."""
        timers_to_display = self._select_timers_to_display(display_max=4)
        display_data = [timer.display_data for timer in timers_to_display]
        if self.platform == MARK_II:
            page = "timer_mark_ii.qml"
        else:
            page = "timer_scalable.qml"
        if timers_to_display:
            self.gui["activeTimers"] = dict(timers=display_data)
            self.gui["activeTimerCount"] = len(timers_to_display)
            self.gui.show_page(page, override_idle=True)
```

Otherwise a blank page is shown continuously cancelling everything in the way
(due to event remove, reinstate if voice in/out; whether you called a timer or not)

File: __init__.py (hunk 1 above)
Comment (chrisveilleux) [reply __init__.py]: Just pushed a few changes regarding the timing of some Mark I faceplate logic.  I think it will address your concerns.

**Comment by chrisveilleux** [comment]
> I think its appropriate to throw some test things in here.
> 
> the "common stop" logic (`def stop`) and the cancel intent are overlapping. "stop" has to be erased from `cancel.voc`, especially since the intent seems to be designed around cancelling non-expired timers.
> This isn't recognizeable if only one unnamed is started and cancelled before/after expiry, but otherwise

The cancel intent requires cancel.voc and timer.voc.  Just saying "stop" will trigger the "common stop" logic while saying "stop timer" will trigger the cancel intent.  I don't understand where you think the overlap is.

**Comment by chrisveilleux** [comment]
> [This](https://github.com/MycroftAI/mycroft-timer/blob/b67d8baf7cfc750e577518f7148322a08031e55c/__init__.py#L707) should be:
> 
> ```python
> def _display_timers_on_gui(self):
>         """Display active timers on a device that supports the QT GUI framework."""
>         timers_to_display = self._select_timers_to_display(display_max=4)
>         display_data = [timer.display_data for timer in timers_to_display]
>         if self.platform == MARK_II:
>             page = "timer_mark_ii.qml"
>         else:
>             page = "timer_scalable.qml"
>         if timers_to_display:
>             self.gui["activeTimers"] = dict(timers=display_data)
>             self.gui["activeTimerCount"] = len(timers_to_display)
>             self.gui.show_page(page, override_idle=True)
> ```
> 
> Otherwise a blank page is shown continuously cancelling everything in the way
> (due to event remove, reinstate if voice in/out; whether you called a timer or not)

Committed this change.  Nice catch.

File: locale/en-us/vocabulary/cancel.voc [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
-clear
-disable
-disabled
-remove
-end
+cancel
+turn off
+kill
+delete
+stop
```
Comment (emphasize) [inline locale/en-us/vocabulary/cancel.voc:5]: This is the overlap i'm talking about

**Review by ken-mycroft — APPROVED** [review]
I stopped after about 20 minutes and clicking view for all the files. This PR is way too long and includes way too many files to be reviewed sanely. I always question the wisdom of completely rewriting a program to fix bugs but for the sake of progress I will add a blanket approval to this skill though I have no expectations that it does not have bugs or that rewriting it is the right thing to do.

**Pull request merged**